### PR TITLE
Autocompletion for zign commands (zo, zoz)

### PR DESCRIPTION
**Pull request by jroimartin**
Show matching files in `dir.zigns` on zo and zoz autocompletion.

### CODE REVIEW

File: libr/core/core.c [hunk 1]
```diff
@@ -628,20 +628,140 @@ static const char *radare_argv[] = {
...
+			iscwd = 1;
+			p++;
+		} else { // *
+			*p = 0;
+			p++;
 		}
+	} else {
+		iscwd = 1;
+		pfree = 1;
+		p = strdup (lpath);
```
Comment (radare) [inline libr/core/core.c]: just p = lpath instead of strdup+free

File: libr/core/core.c [hunk 2]
```diff
@@ -628,20 +628,142 @@ static const char *radare_argv[] = {
...
+		} else {
+			list = r_sys_dir ("/");
+		}
+	} else {
+		if (*lpath=='~') { // if implicit home
+			char *lala = r_str_home (lpath + 1);
+			free (lpath);
+			lpath = lala;
+		} else if (*lpath!='.' && *lpath!='/') { // ifnot@home
+			char *o = malloc (strlen (lpath) + 4);
```
Comment (radare) [inline libr/core/core.c]: check if null

File: libr/core/core.c [hunk 3]
```diff
@@ -628,20 +628,142 @@ static const char *radare_argv[] = {
...
+	} else if (isroot) {
+		const char *lastslash = r_str_lchr (lpath, '/');
+		if (lastslash && lastslash[1]) {
+			list = r_sys_dir (lpath);
+		} else {
+			list = r_sys_dir ("/");
+		}
+	} else {
+		if (*lpath=='~') { // if implicit home
+			char *lala = r_str_home (lpath + 1);
```
Comment (radare) [inline libr/core/core.c]: check for null

File: libr/core/core.c [hunk 4]
```diff
@@ -628,20 +628,142 @@ static const char *radare_argv[] = {
...
+			free (lpath);
+			lpath = lala;
+		} else if (*lpath!='.' && *lpath!='/') { // ifnot@home
+			char *o = malloc (strlen (lpath) + 4);
+			memcpy (o, "./", 2);
+			p = o+2;
```
Comment (radare) [inline libr/core/core.c]: missing spaces around +

File: libr/core/core.c [hunk 5]
```diff
@@ -628,20 +628,142 @@ static const char *radare_argv[] = {
...
+
+	for (i = 0; paths[i]; i ++) {
+		char *buf = r_str_newf ("%s%s%s", paths[i], R_SYS_DIR, tinput);
+		if (!buf) {
+			break;
+		}
+		argv_idx += autocompleteProcessPath (line, buf, argv_idx);
+		free (buf);
+	}
+
```
Comment (radare) [inline libr/core/core.c]: empty line

File: libr/core/core.c [hunk 6]
```diff
@@ -628,20 +628,142 @@ static const char *radare_argv[] = {
...
+		free (lpath);
+		lpath = strdup (".");
 	}
-	return 0;
+	if (pfree) {
+		if (p) {
+			if (*p) {
+				n = strlen (p);
+			} else {
+				free (p);
```
Comment (radare) [inline libr/core/core.c]: maybe just *p=0

Comment (radare) [inline libr/core/core.c]: and avoid the unnecessary allocations

File: libr/core/core.c [hunk 7]
```diff
@@ -628,20 +628,142 @@ static const char *radare_argv[] = {
...
+				p = strdup ("");
+			}
+		}
+	} else {
+		if (p) { if (*p) n = strlen (p); else p = ""; }
```
Comment (radare) [inline libr/core/core.c]: expand into multiple lines and add braces

File: libr/core/core.c [hunk 8]
```diff
@@ -831,134 +957,45 @@ static int autocomplete(RLine *line) {
...
-			} else {
-				eprintf ("\nInvalid directory (%s)\n", path);
-			}
-			tmp_argv[i] = NULL;
-			line->completion.argc = i;
-			line->completion.argv = tmp_argv;
-			free (path);
-			if (pfree)
-				free (p);
+			char *paths[2] = { zignpath, NULL };
```
Comment (radare) [inline libr/core/core.c]: we can avoid going into autocompleteFilename and free if zignpath is null, not sure if we are shortening the path in there too

Comment (jroimartin) [inline libr/core/core.c]: what do you mean? If zignpath is null you still have to autocomplete current dir

**Comment by jroimartin** [comment]
All the comments are related to `autocompleteProcessPath`, I didn't touch that code (I only moved it to a separate function) because it must be rewritten. It's buggy and, based on the comment, has mem leaks.

File: libr/core/core.c (hunk 5 above)
Comment (jroimartin) [inline libr/core/core.c]: intended.

File: libr/core/core.c (hunk 8 above)
Comment (jroimartin) [inline libr/core/core.c]: Anyway, I changed a bit the code in 2d243d3 to make it more clear.

**Pull request merged**